### ISSUE TITLE
Realm: Deferred allocation breaks under a tight memory budget

### ISSUE THREAD

**Issue by magnatelee** [StanfordLegion/legion#902]
### reproducer.rg

```
import "regent"
import "bishop"

mapper
task#f region#r {
  create: demand;
}
end

task f(r : region(ispace(int1d), int)) where reads writes(r) do end

task main()
  var size = 16777216
  var r = region(ispace(int1d, size), int) -- 64MB
  var s = region(ispace(int1d, size), int) -- 64MB
  for i = 0, 10 do
    f(r)
    f(s)
  end
end

regentlib.start(main, bishoplib.make_entry())
```

### Command line
```
regent.py reproducer.rg -lg:window 10 -ll:csize 256 -level 5
```

### Symptom
```
$ regent.py reproducer.rg -lg:window 10 -ll:csize 256 -level 5
terra: /gpfs/fs1/wonchanl/legate.core/legion/runtime/realm/mem_impl.cc:787: void Realm::MemoryImpl::deferred_destruction_triggered(Realm::RegionInstanceImpl*, bool, Realm::TimeLimit): Assertion `future_allocator.lookup(it2->inst->me, f_first, f_size) && (f_first == offset) && (f_size == it2->bytes)' failed.
```

**Comment by streichler** [comment]
I believe this is fixed with commit 41aeaa3.  @magnatelee can you poke at it some more and see if you agree?

**Comment by magnatelee** [comment]
I threw several different flags to the test and I wasn't able to make it break, so I'd say the issue is fixed.
